File: packages/agentenkit/test/subagent-slots.test.ts
import { describe, expect, it } from 'bun:test';
import { simulateReadableStream } from 'ai';
import type { LanguageModelV1StreamPart } from '@ai-sdk/provider';
import { MockLanguageModelV1 } from 'ai/test';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryStorage } from '../src/adapters/memory.js';
import { RunSlots } from '../src/core/subagent.js';
import { resolveConfig, type AgentConfig } from '../src/core/types.js';

// Workstream F: the subagent cap is per run and per depth, a wait for a slot
// ends on a stop, and a child that was stopped or cut short is never taken
// for finished. The same cases run in the Go package
// (subagent_slots_test.go).

interface Step {
  text?: string;
  calls?: Array<{ id: string; instructions: string }>;
  delayMs?: number;
  noFinish?: boolean;
}

/** Answers by who is asking rather than by call order: nested runs call it in
 *  parallel, so their order is not fixed. */
function routedModel(route: (brief: string, answered: boolean) => Step) {
  return new MockLanguageModelV1({
    provider: 'mock',
    modelId: 'mock-routed',
    doStream: async ({ prompt, abortSignal }: any) => {
      const first = (prompt as any[]).find((m) => m.role === 'user');
      const brief = (first?.content ?? []).map((p: any) => p.text ?? '').join('');
      const answered = (prompt as any[]).some((m) => m.role === 'tool');
      const s = route(brief, answered);
      if (s.delayMs) {
        await new Promise<void>((resolve, reject) => {
          const timer = setTimeout(resolve, s.delayMs);
          abortSignal?.addEventListener('abort', () => {
            clearTimeout(timer);
            reject(Object.assign(new Error('aborted'), { name: 'AbortError' }));
          });
        });
      }
      const chunks: LanguageModelV1StreamPart[] = [];
      if (s.text) chunks.push({ type: 'text-delta', textDelta: s.text });
      for (const c of s.calls ?? []) {
        chunks.push({
          type: 'tool-call', toolCallType: 'function', toolCallId: c.id, toolName: 'spawnSubagent',
          args: JSON.stringify({ name: 'helper', instructions: c.instructions }),
        });
      }
      if (!s.noFinish) {
        chunks.push({
          type: 'finish',
          finishReason: s.calls?.length ? 'tool-calls' : 'stop',
          usage: { promptTokens: 10, completionTokens: 5 },
        });
      }
      return { stream: simulateReadableStream({ chunks }), rawCall: { rawPrompt: null, rawSettings: {} } };
    },
  });
}

async function routedRuntime(route: (brief: string, answered: boolean) => Step, config: Partial<AgentConfig> = {}) {
  const model = routedModel(route);
  const storage = new MemoryStorage();
  const bus = new MemoryBus();
  const queue = new MemoryQueue();
  const admin = new MemoryAdminStore();
  const runtime = await setupAgentCore({
    storage, bus, queue, kv: new MemoryKv(), admin,
    resolveModel: () => ({ instance: () => model, contextWindow: 128_000 }),
    config: resolveConfig({ stopPollMs: 5, runRetryBackoffMs: 1, ...config }),
  });
  const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o', subagents: true });
  const events = (threadId: string, type: string) => bus.published.filter((e) => e.threadId === threadId && e.type === type);
  const state = async (threadId: string) => (await storage.threads.get(threadId))!.state;
  return { runtime, chat, queue, admin, events, state };
}

describe('the subagent cap (§2.7)', () => {
  it('nested spawns at the default limit do not deadlock', async () => {
    const r = await routedRuntime((brief, answered) => {
      if (brief === 'go' && !answered) {
        // The parent fills every slot at depth 1.
        return { calls: [{ id: 's1', instructions: 'child' }, { id: 's2', instructions: 'child' }, { id: 's3', instructions: 'child' }] };
      }
      if (brief === 'go') return { text: 'parent done' };
      // Each child waits a moment before it spawns, so all three hold their
      // slots before any grandchild asks for one.
      if (brief === 'child' && !answered) return { delayMs: 50, calls: [{ id: 'g', instructions: 'grandchild' }] };
      if (brief === 'child') return { text: 'child done' };
      return { text: 'grandchild done' };
    });
    const ran = await r.chat.run({ prompt: 'go' });
    const finished = r.runtime.worker.handleJob(r.queue.items.shift()!);
    const outcome = await Promise.race([
      finished.then(() => 'done'),
      new Promise((resolve) => setTimeout(() => resolve('deadlocked'), 5_000)),
    ]);
    expect(outcome).toBe('done');
    expect(await r.state(ran.threadId)).toBe('COMPLETED');
    expect(r.events(ran.threadId, 'SUBAGENT_COMPLETED')).toHaveLength(6); // 3 children, 3 grandchildren
  }, 10_000);

  it('slots are per depth', async () => {
    const slots = new RunSlots(1);
    const release = await slots.acquire(1);
    const deeper = await Promise.race([
      slots.acquire(2).then(() => 'got it'),
      new Promise((resolve) => setTimeout(() => resolve('blocked'), 100)),
    ]);
    expect(deeper).toBe('got it');
    release();
  });

  it('a wait for a slot ends when the run is stopped', async () => {
    const slots = new RunSlots(1);
    const release = await slots.acquire(1);
    const stop = new AbortController();
    setTimeout(() => stop.abort(new Error('stopped')), 20);
    await expect(slots.acquire(1, stop.signal)).rejects.toThrow('stopped');
    release();
  });
});

describe('a child that did not finish (§2.7)', () => {
  it('a child cut off by a stop is recorded cancelled', async () => {
    let childStarted = false;
    const r = await routedRuntime((brief, answered) => {
      if (brief === 'go' && !answered) return { calls: [{ id: 's1', instructions: 'child' }] };
      if (brief === 'go') return { text: 'parent done' };
      childStarted = true;
      return { text: 'never', delayMs: 2_000 };
    });
    const ran = await r.chat.run({ prompt: 'go' });
    const running = r.runtime.worker.handleJob(r.queue.items.shift()!);
    while (!childStarted) await new Promise((resolve) => setTimeout(resolve, 5));
    await r.chat.stop(ran.threadId);
    await running;
    expect(r.events(ran.threadId, 'SUBAGENT_COMPLETED')).toHaveLength(0);
    const childId = (r.events(ran.threadId, 'SUBAGENT_STARTED')[0]!.payload as any).agentId;
    expect((await r.admin.runs.get(childId))!.state).toBe('CANCELLED');
  });

  it('a child whose stream ends without a finish fails', async () => {
    const r = await routedRuntime((brief, answered) => {
      if (brief === 'go' && !answered) return { calls: [{ id: 's1', instructions: 'child' }] };
      if (brief === 'go') return { text: 'parent done' };
      return { text: 'half an answer', noFinish: true };
    });
    const ran = await r.chat.run({ prompt: 'go' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    expect(await r.state(ran.threadId)).toBe('COMPLETED'); // the parent goes on
    const failed = r.events(ran.threadId, 'SUBAGENT_FAILED');
    expect(failed).toHaveLength(1);
    expect((failed[0]!.payload as any).error).toContain('without a finish');
    expect(r.events(ran.threadId, 'SUBAGENT_COMPLETED')).toHaveLength(0); // partial text is not a result
  });
});

describe('the main agent (§2.8)', () => {
  it('a stream cut without a finish is not completed', async () => {
    let calls = 0;
    const r = await routedRuntime(() => (calls++ === 0 ? { text: 'part', noFinish: true } : { text: 'whole' }));
    const ran = await r.chat.run({ prompt: 'hi' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    expect(await r.state(ran.threadId)).not.toBe('COMPLETED'); // a step that never finished
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    expect(await r.state(ran.threadId)).toBe('COMPLETED'); // the retry completes it
  });
});

File: packages/go-agentenkit/subagent_slots_test.go
package agentenkit_test

import (
	"context"
	"errors"
	"strings"
	"sync/atomic"
	"testing"
	"time"

	"github.com/zendev-sh/goai/provider"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/core"
)

// Workstream F: the subagent cap is per run and per depth, a wait for a slot
// ends on a stop, and a child that was stopped or cut short is never taken
// for finished. The same cases run in the TS package
// (test/subagent-slots.test.ts).

// routedModel answers by who is asking rather than by call order: nested
// runs call it in parallel, so their order is not fixed.
type routedModel struct {
	route func(brief string, answered bool) step
}

func (m *routedModel) ModelID() string { return "mock-routed" }

func (m *routedModel) pick(p provider.GenerateParams) step {
	brief, answered := "", false
	for _, msg := range p.Messages {
		for _, part := range msg.Content {
			if msg.Role == provider.RoleUser && brief == "" && part.Type == provider.PartText {
				brief = part.Text
			}
			if part.Type == provider.PartToolResult {
				answered = true
			}
		}
	}
	return m.route(brief, answered)
}

func (m *routedModel) DoGenerate(ctx context.Context, p provider.GenerateParams) (*provider.GenerateResult, error) {
	return scripted(m.pick(p)).DoGenerate(ctx, p)
}

func (m *routedModel) DoStream(ctx context.Context, p provider.GenerateParams) (*provider.StreamResult, error) {
	return scripted(m.pick(p)).DoStream(ctx, p)
}

func routedRuntime(t *testing.T, route func(brief string, answered bool) step, tune ...func(*agentenkit.AgentConfig)) (*harness, *agentenkit.AgentHandle) {
	model := &routedModel{route: route}
	h := makeRuntimeOpts(t, scripted(step{text: "unused"}), func(o *agentenkit.RuntimeOptions) {
		o.ResolveModel = func(string) (agentenkit.ResolvedModel, error) {
			return agentenkit.ResolvedModel{Instance: func() provider.LanguageModel { return model }, ContextWindow: 128_000}, nil
		}
	}, tune...)
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{
		Name: "chat", Model: "gpt-4o", Subagents: &agentenkit.SubagentsConfig{},
	})
	return h, chat
}

func spawn(id, instructions string) call {
	return call{id, "spawnSubagent", `{"name":"helper","instructions":"` + instructions + `"}`}
}

func TestSubagents_NestedSpawnsAtTheDefaultLimitDoNotDeadlock(t *testing.T) {
	h, chat := routedRuntime(t, func(brief string, answered bool) step {
		switch {
		case brief == "go" && !answered: // the parent fills every slot at depth 1
			return step{calls: []call{spawn("s1", "child"), spawn("s2", "child"), spawn("s3", "child")}}
		case brief == "go":
			return step{text: "parent done"}
		case brief == "child" && !answered:
			// Each child spawns one level deeper, after a moment, so all
			// three hold their slots before any grandchild asks for one.
			return step{delay: 50 * time.Millisecond, calls: []call{spawn("g", "grandchild")}}
		case brief == "child":
			return step{text: "child done"}
		default:
			return step{text: "grandchild done"}
		}
	})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "go"})
	done := make(chan struct{})
	go func() { defer close(done); h.handleNext(t) }()
	select {
	case <-done:
	case <-time.After(5 * time.Second):
		t.Fatal("the run deadlocked on the subagent cap")
	}
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateCompleted, "completed")
	mustEqual(t, len(h.events(ran.ThreadID, "SUBAGENT_COMPLETED")), 6, "three children, three grandchildren")
}

func TestSubagents_SlotsArePerDepth(t *testing.T) {
	slots := core.NewRunSlots(1)
	release, err := slots.Acquire(context.Background(), 1)
	if err != nil {
		t.Fatal(err)
	}
	defer release()
	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()
	if _, err := slots.Acquire(ctx, 2); err != nil {
		t.Fatalf("a deeper level has slots of its own: %v", err)
	}
}

func TestSubagents_AWaitForASlotEndsWhenTheRunIsStopped(t *testing.T) {
	slots := core.NewRunSlots(1)
	release, _ := slots.Acquire(context.Background(), 1)
	defer release()
	ctx, cancel := context.WithCancel(context.Background())
	go func() { time.Sleep(20 * time.Millisecond); cancel() }()
	if _, err := slots.Acquire(ctx, 1); !errors.Is(err, context.Canceled) {
		t.Fatalf("the wait gives up on the stop: %v", err)
	}
}

func TestSubagents_AChildCutOffByAStopIsRecordedCancelled(t *testing.T) {
	var childStarted atomic.Bool
	h, chat := routedRuntime(t, func(brief string, answered bool) step {
		switch {
		case brief == "go" && !answered:
			return step{calls: []call{spawn("s1", "child")}}
		case brief == "go":
			return step{text: "parent done"}
		default:
			childStarted.Store(true)
			return step{text: "never", delay: 2 * time.Second}
		}
	}, func(c *agentenkit.AgentConfig) { c.StopPoll = 5 * time.Millisecond })
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "go"})
	job, _ := h.queue.Shift()
	done := make(chan struct{})
	go func() { defer close(done); _, _ = h.rt.Worker.HandleJob(h.ctx, job) }()
	waitFor(t, childStarted.Load)
	if _, err := chat.Stop(h.ctx, ran.ThreadID, nil); err != nil {
		t.Fatal(err)
	}
	<-done
	mustEqual(t, len(h.events(ran.ThreadID, "SUBAGENT_COMPLETED")), 0, "never completed")
	childID := payload(h.events(ran.ThreadID, "SUBAGENT_STARTED")[0])["agentId"].(string)
	rec, _ := h.admin.Runs().Get(h.ctx, childID)
	mustEqual(t, rec.State, agentenkit.StateCancelled, "the child is recorded cancelled")
}

func TestSubagents_AChildWhoseStreamEndsWithoutAFinishFails(t *testing.T) {
	h, chat := routedRuntime(t, func(brief string, answered bool) step {
		switch {
		case brief == "go" && !answered:
			return step{calls: []call{spawn("s1", "child")}}
		case brief == "go":
			return step{text: "parent done"}
		default:
			return step{text: "half an answer", noFinish: true}
		}
	})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "go"})
	h.handleNext(t)
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateCompleted, "the parent goes on")
	failed := h.events(ran.ThreadID, "SUBAGENT_FAILED")
	mustEqual(t, len(failed), 1, "the child failed")
	if !strings.Contains(payload(failed[0])["error"].(string), "without a finish") {
		t.Fatalf("with why: %v", payload(failed[0]))
	}
	mustEqual(t, len(h.events(ran.ThreadID, "SUBAGENT_COMPLETED")), 0, "its partial text is not a result")
}
